src/bepasty/tests/test_rest_server.py

```diff
@@ -0,0 +1,817 @@
+#
+# REST api tests
+#
+
+import os
+import base64
+import threading
+import time
+import copy
+import hashlib
+import re
+from requests.auth import _basic_auth_str
+from flask import current_app
+from flask import url_for
+
+import pytest
+
+from ..app import create_app
+from ..config import Config
+from ..constants import FILENAME, TYPE, LOCKED, SIZE, COMPLETE, HASH, \
+    TIMESTAMP_DOWNLOAD, TIMESTAMP_UPLOAD, TIMESTAMP_MAX_LIFE, TRANSACTION_ID
+from ..utils.date_funcs import time_unit_to_sec
+
+UPLOAD_DATA = b"""\
+#!/usr/bin/python3
+
+print('hello, world')
+"""
+
+
+class FakeTime:
+    """Overwrite time.time to control the timestamp that is used in server"""
+
+    def __init__(self, now=None):
+        self.orig = time.time
+        self.now = now
+
+    def set_time(self, now=None):
+        self.now = now
+        if now and time.time != self.get_time:
+            time.time = self.get_time
+        elif now is None and time.time == self.get_time:
+            time.time = self.orig
+
+    def get_time(self):
+        return self.now
+
+    def __enter__(self):
+        self.set_time(self.now)
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.set_time(None)
+
+
+def wait_background():
+    """Wait until background threads terminate."""
+
+    main_thread = threading.current_thread()
+    for t in threading.enumerate():
+        if t is not main_thread:
+            t.join()
+
+
+@pytest.fixture
+def client_fixture(tmp_path):
+    with FakeTime() as faketime:
+        Config.PERMISSIONS = {
+            'admin': 'admin,list,create,read,delete',
+            'full': 'list,create,read,delete',
+            'none': '',
+        }
+        Config.STORAGE_FILESYSTEM_DIRECTORY = str(tmp_path)
+
+        app = create_app()
+        app.config['TESTING'] = True
+
+        with app.test_client() as client:
+            with app.app_context():
+                yield app, client, faketime
+
+        wait_background()
+
+
+class TmpConfig:
+    def __init__(self, app, tmp_config=None):
+        self.config = app.config
+        self.orig_config = copy.deepcopy(app.config)
+        self.tmp_config = {} if tmp_config is None else tmp_config
+
+    def __enter__(self):
+        self.config.update(self.tmp_config)
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.config.update(self.orig_config)
+
+
+class RestUrl:
+    def __init__(self, fid=None):
+        self.fid = fid
```

Comment:
minor nitpick: `fid` is not a very speaking name. bepasty usually talks about items, so how about `item_id`?

Reply:
i wanted to use shorter name though. changed to item_id